In [45]:
import validaciones

# Ejercicios Python

## Zen of Python

Lee una vez mas el Zen of Python antes de iniciar con la tarea

In [46]:
import this

## 1 - XOR

En la siguiente celda crea una funcion para la compuerta logica XOR 

<img src="assets/xor.png" width="400">


In [47]:
def xor(a,b):
    """
    Esta funcion recibe dos valores booleanos, ejemplo xor(1, 0) o True y False
    y retorna 0 o 1 como resultado de la operacion logica XOR
    """
    return int((a or b) & ((not a) | (not b)))

In [48]:
xor(0 , 1) 

1

In [49]:
xor(0 , 0)

0

In [50]:
xor(True , False)

1

### Pruebas  Automaticas

Todas las pruebas deben funcionar para obetener los puntos del ejercicio.

In [51]:
validaciones.xor_function(xor)

Prueba 1
Prueba 1 [CORRECTA]
Prueba 2
Prueba 2 [CORRECTA]


## 2 - Sumando N series

Tenemos una secuencia cuyo $n^{th}$ es:
\begin{equation*}
\ T_n = n^2 - (n-1)^2
\end{equation*}

Hay que evaluar las series:
\begin{equation*}
\ S_n = T_1 + T_2 + ... + T_n
\end{equation*}


**Instrucciones**
En la siguiente celda en la funcion summingSeries debemos **retornar** en valor de $S_n$, tenemos como entrada n

**Ejemplo 1:**

Si la entrada es 2

Salida:
4

Explicacion:

$T_1 = 1$

$T_2 = 3$

$S_2 = T_1 + T_2$

$S_2 = 4 $



**Tip :**
Antes te implementar $S_n$ la solución, analicemos algebraicamente el problema para encontrar una solución equivalente optima.


In [52]:
def summatoriaSerie(n):
    return n**2

    """
    Luego de tratar de resolverlo con recursión, la validación lanza un error en la prueba 5 y 6. Pero podemos simplificar la
    la expresión T y la sumatoria S como el cuadrado del número N, es por eso que el return de la función es n**2

    if n == 1:
        return 1
    else:
        return (2*n - 1) + summatoriaSerie(n - 1);
    """

#### 2.1 - Ciclos

Crea un ciclo for que imprima el resultado de la funcion **summingSeries** desde 1 hasta 10

In [53]:
for i in range(1,11):
    print(summatoriaSerie(i))

1
4
9
16
25
36
49
64
81
100


### Pruebas  Automaticas

Todas las pruebas deben funcionar para obetener los puntos del ejercicio.

##### Importante :
En este ejercicio se evalua que la funcion sea optima, el tiempo de ejecucion tiene que ser el menor posible, se va a evaluar el resultado correcto y el tiempo de ejecucion, si en alguna de las pruebas obtienes **[FALLO: TIMEOUT]** la funcion no es optima debes cambiarla hasta que todas las pruebas pasen

In [54]:
validaciones.summingSeries_function(summatoriaSerie)

Prueba 1
Prueba 1[CORRECTA]
Prueba 2
Prueba 2[CORRECTA]
Prueba 3
Prueba 3[CORRECTA]
Prueba 4
Prueba 4[CORRECTA]
Prueba 5
Prueba 5[CORRECTA]
Prueba 6
Prueba 6[CORRECTA]


## 3 - Factores Primos

Para cada $n$ queremos obtener el conteo maximo de numeros primos unicos en en rango inclusivo $[1, n]$,
y retornar el valor del conteo en una nueva linea

**Nota:** Recuerda que un numero primo solo es divisible por el mismo y que 1 no es un numero primo

##### Ejemplos:

###### Ejemplo 1
Entrada: 1

Salida esperada: 1

Explicacion: El numero maximo de factores primos unicos en el rango inclusivo $[1,1]$ es $0$, porque $1$ no es un numero primo.

###### Ejemplo 2
Entrada: 3

Salida esperada: 1

Explicacion: El numero maximo de factores primos unicos en el rango inclusivo $[1,3]$ es $1$, porque el numero $3$ tiene 1 factor primo unico (el mismo) 

###### Ejemplo 3
Entrada: 500

Salida esperada: 4

Explicacion: El numero maximo de factores primos unicos en el rango inclusivo $[1,500]$ es $4$, porque el producto de los primeros cuatro numeros primos unicos es $2 \times 3 \times 5 \times 7 = 210$ y no hay ningun numero primo unico que multiplicado por el resultado sea $	\leqslant 500$



#### Tip: 
Utiliza la funcion `range()` de python


In [55]:
# Calcular los factores
def conteoPrimos(n):
    contador = 0
    multiplicador = 1
    
    if n == 1:
        return 0
    elif n == 2:
        return 1

    for i in range(2,n):
        if esPrimo(i) == True:
            multiplicador *= i
            if multiplicador <= n:
                contador+=1
            else:
                break
    return contador;

# Calcular si un número es primo
# Sabiendo que:
# -Un número primo solo es divisible exactamente por él mismo y la unidad
# -El uno no es un número primo
# -El dos es el primer número primo

def esPrimo(x):
    es_primo = True

    if x == 1:
        return not es_primo
    elif x == 2:
        return es_primo
    
    for i in range(2,x):
        if (x % i) == 0:
            es_primo = False
            break;
    return es_primo;


### Pruebas

In [56]:
conteoPrimos(1)

0

In [57]:
conteoPrimos(500)

4

### Pruebas  Automaticas

Todas las pruebas deben funcionar para obetener los puntos del ejercicio.

##### Importante :
En este ejercicio se evalua que la funcion sea optima, el tiempo de ejecucion tiene que ser el menor posible, se va a evaluar el resultado correcto y el tiempo de ejecucion, si en alguna de las pruebas obtienes **[FALLO: TIMEOUT]** la funcion no es optima debes cambiarla hasta que todas las pruebas pasen

In [58]:
validaciones.primeCount_function(conteoPrimos)

Prueba 1
Prueba 1[CORRECTA]
Prueba 2
Prueba 2[CORRECTA]
Prueba 3
Prueba 3[CORRECTA]
Prueba 4
Prueba 4[CORRECTA]
Prueba 5
Prueba 5[CORRECTA]
Prueba 6
Prueba 6[CORRECTA]
Prueba 7
Prueba 7[CORRECTA]
Prueba 8
Prueba 8[CORRECTA]
Prueba 9
Prueba 9[CORRECTA]
Prueba 10
Prueba 10[CORRECTA]
Prueba 11
Prueba 11[CORRECTA]
Prueba 12
Prueba 12[CORRECTA]


## 4 - Metodo de Euler

El metodo de Euler se utiliza para aproximar la solucion particular de una ecuacion diferencial dado un valor inicial. Con esta informacion se sabe que la grafica de esa solucion pasa a traves del punto $(x_0,y_0)$ y tiene una pendiente $F(x_0,y_0)$ en ese punto, esto da un punto de partida para aproximar la solucion.

A partir del punto inicial, se sigue en la direccion indicada por la pendiente. Mediante un pequeño paso $h$, se mueve a lo largo de la recta tangente hasta llegar al punto $(x_1,y_1)$ donde:
\begin{equation*}
\ x_1 = x_0 + h       \\
\ y_1 = y_0 + hF(x_0,y_0)\\
\end{equation*}


<img src="assets/Euler.png" width="350">

Como se muestra en la figura se considera $(x_1,y_1)$ como un nuevo punto inicial.

Los valores de $x$ son:
\begin{equation*}
\ x_1 = x_0 + h       \\
\ x_2 = x_1 + h       \\
\    .                \\
\    .                \\
\    .                \\
\ x_n = x_{n-1} + h   \\
\end{equation*}


Los valores de $y$ son:
\begin{equation*}
\ y_1 = y_0 + hF(x_0,y_0)\\
\ y_1 = y_1 + hF(x_1,y_1)\\
\       . \\
\       . \\
\       . \\
\ y_n = y_{n-1} + hF(x_{n-1},y_{n-1})\\
\end{equation*}


#### Problema 4
El ejecicio consiste en implementar la funcion *euler_metod* esta funcion recibe como parametro el 
* punto inicial $x_0$
* punto inicial $y_0$
* Paso $h$ 
* numero de iteraciones $n$ 
* La equacion diferencial que queremos aproximar, esta funcion tambien recibe los parametros $x$ y $y$
* La funcion de la solucion exacta que tambien tambien recibe los parametros $x$ y $y$

La funcion debe retornar el valor de $y_n$ y el error
$error=|y - y_n|$ 

In [59]:
import math

In [60]:
# La función imprime los resultados necesarios para completar la tabla en tipo Markdown

def euler_method(x,y,h,n, diff_eq, func):
    for i in range(0,n+1):
        print(round(x,1), round(func(x),4), round(y,4), round(abs(func(x) - y),4))
        y = y + h*(diff_eq(x, y))
        x = x + h


### Ejemplo

Queremos aproximar la solucion particular de la ecuacion diferencial
\begin{equation*}
\ \frac{\partial y}{\partial x} = -2y \\
\end{equation*}
Donde $y(0) = 4$ vamos a unsa un de $h=0.1$

Tambien vamos a calcular el error en este caso podemos verificar el error  con la solucion exacta:
\begin{equation*}
\ y = 4e^{-2x} \\
\end{equation*}

In [61]:
### Implementando las funciones del ejemplo
def y_prime(x, y):
    return -2*y

def y(x):
    return 4 * math.exp(-2*x)

In [62]:
"""
Llamamos a la funcion de euler
euler_method(0,4,0.1,10, y_prime,y)
Parametros :
x = 0
y = 4
d = 0.1
n = 10
diff_eq = y_prime
func = y
"""

print("Resultados en orden: X, Y, Yn, Error\n")
euler_method(0,4,0.1,10, y_prime,y)

Resultados en orden: X, Y, Yn, Error

0 4.0 4 0.0
0.1 3.2749 3.2 0.0749
0.2 2.6813 2.56 0.1213
0.3 2.1952 2.048 0.1472
0.4 1.7973 1.6384 0.1589
0.5 1.4715 1.3107 0.1608
0.6 1.2048 1.0486 0.1562
0.7 0.9864 0.8389 0.1475
0.8 0.8076 0.6711 0.1365
0.9 0.6612 0.5369 0.1243
1.0 0.5413 0.4295 0.1118


### Solucion:

Utiliza esta tabla como guia para validar que tu funcion este correcta,
y reemplaza los valores que faltan con la salida de tu funcion.

| x     | 0 | 0.1         | 0.2    | 0.4    | 0.6    | 0.8    | 1            |
|-------|---|-------------|--------|--------|--------|--------|--------------|
| y     | 4 | 3.2749      | 2.6813 | 1.7973 | 1.2048 | 0.8076 | 0.5413       |
| $y_n$ | 4 | 3.2         | 2.56   | 1.6384 | 1.0486 | 0.6711 | 0.4295       |
| error | 0 | 0.0749      | 0.1213 | 0.1589 | 0.1562 | 0.1365 | 0.1118       |

### 4.1 Probando la implementacion con una aplicación real de data science

Podemos utilizar ecuaciones diferenciales  como modelos matemáticos de fenónemos reales, por ejemplo si llegas a trabajar como data scientist para una farmaceutica  puede que interese analizar la desintegración radioactiva de cierta sustancia y como esto cambia en el tiempo
#### Desintegracion Radiactiva

$A$ es la cantidad de sustancia en el tiempo $t$, la derivada $\frac{\partial A}{\partial t}$ es rapidez a la que se desintegran los nucleos de una sustancia.

Queremos aproximar la solucion particular de la ecuacion diferencial
\begin{equation*}
\ \frac{\partial A}{\partial t} =  -0.56A \\
\end{equation*}
Donde $A(0) = 1$ vamos a un tamaño de $h=0.1$

Tambien vamos a calcular el error en este caso podemos verificar el error  con la solucion exacta:
\begin{equation*}
\ A(t) = e^{-0.56t} \\
\end{equation*}

**Nota** En este caso sencillo, poseemos la solución exacta a la ED por lo cual no se necesita una solución aproximada , pero en aplicaciones y proyectos reales muchas veces la solución exacta no se posee, o es muy difícil de calcular y recurrimos a soluciones aproximadas por computadora , para este ejercicio el saber la solución exacta nos sirve para medir la exactitud de nuestra solución aproximada.

In [63]:
def A_prime(t, A):
    return -0.56*A

def A(t):
    return math.exp(-0.56*t)

In [64]:
print("Resultados en orden: X, Y, Yn, Error\n")
euler_method(0,1,0.1,10, A_prime,A)

Resultados en orden: X, Y, Yn, Error

0 1.0 1 0.0
0.1 0.9455 0.944 0.0015
0.2 0.894 0.8911 0.0029
0.3 0.8454 0.8412 0.0041
0.4 0.7993 0.7941 0.0052
0.5 0.7558 0.7497 0.0061
0.6 0.7146 0.7077 0.007
0.7 0.6757 0.668 0.0077
0.8 0.6389 0.6306 0.0083
0.9 0.6041 0.5953 0.0088
1.0 0.5712 0.562 0.0092


### Solucion 4.1:

Tabla de resultados para el ejercicio 4.1 de Desintegracion Radiactiva

| t     | 0 | 0.1         | 0.2    | 0.4    | 0.6    | 0.8    | 1            |
|-------|---|-------------|--------|--------|--------|--------|--------------|
| a     | 1 | 0.9455      | 0.894  | 0.7993 | 0.7146 | 0.6389 | 0.5712       |
| $y_n$ | 1 | 0.944       | 0.8911 | 0.7941 | 0.7077 | 0.6306 | 0.562        |
| error | 0 | 0.0015      | 0.0029 | 0.0052 | 0.007  | 0.0083 | 0.0092       |

## 5 - Rotacion de listas hacia la izquierda

Una rotacion a la izquierda en una lista mueve cada elemento de la lista una vez hacia la izquierda. Por ejemplo, si hacemos $1$ rotacion en en el arreglo $[1,2,3,4,5]$, el arreglo resultante seria $[2,3,4,5,1]$

#### Descipcion de la funcion
Dado un arreglo $a$ de $n$ integers y un numero $d$ de rotaciones, realizar $d$ rotaciones hacia la izquierda. La funcion debe retorner el arreglo re-ordenado.

**rotIzquierda** tiene los siguientes parametros:

* $a$ Un arreglo de numeros entreros.
* $d$ numero de rotaciones.

In [65]:
def rotIzquierda(a, d):
    movimientos_reales = d%len(a)

    slice_movimientos = slice(movimientos_reales)
    a.extend(a[slice_movimientos])
    del a[0:movimientos_reales]

    return a

In [66]:
rotIzquierda([1, 2, 3, 4, 5], 4)
# Resultado: [5, 1, 2, 3, 4]

[5, 1, 2, 3, 4]

### Pruebas  Automaticas

Todas las pruebas deben funcionar para obetener los puntos del ejercicio.

##### Importante :
En este ejercicio se evalua que la funcion sea optima, el tiempo de ejecucion tiene que ser el menor posible, se va a evaluar el resultado correcto y el tiempo de ejecucion, si en alguna de las pruebas obtienes **[FALLO: TIMEOUT]** la funcion no es optima debes cambiarla hasta que todas las pruebas pasen

#### TIP:
Utiliza **slice** en el array para optimizar la funcion

In [67]:
validaciones.rotIzquierda_function(rotIzquierda)

Prueba 1
Prueba 1[CORRECTA]
Prueba 2
Prueba 2[CORRECTA]
Prueba 3
Prueba 3[CORRECTA]
